File: homework_06_09_22/homework.cpp
#include <iostream>
#include <vector>
#include <map>
//1.
int binarySearchIterative(std::vector<int>& arr, int target);
int binarySearchRecursive(std::vector<int>& arr, int target, int begin, int end);

//2.
int _strlen(char* str);

//3.
void reverse(char* str);

//4.
void to_upper_case(char* str);

//5.
int _add(int op1, int op2);
int _sub(int op1, int op2);
int _mul(int op1, int op2);
int _div(int op1, int op2);

//6.
char* _strcat(char* s1, char* s2);

//7.
void _strset(char* s, char ch);

//8.
void _strmove(char* s, int n);



int main() {
        //int target;
        //int begin = 0;
        //int end;
        //std::vector<int> arr = {1,2,3,4,5,6,7,8,9,10};
        //end = arr.size() - 1;
        //std::cin >> target;

        //1.
        //std::cout << binarySearchIterative(arr, target) << std::endl;
        //std::cout << binarySearchRecursive(arr, target, begin, end) << std::endl;

        //2.
        /*
        char* str = new char[100];
        std::cin.getline(str, 100);
        std::cout << _strlen(str) << std::endl;
        //char ch;
        //std::cin >> ch;
        //_strset(str, ch);

        //3.
        //reverse(str);

        //4.
        //to_upper_case(str);

        //5.
        int op1;
        int op2;
        char ins;
        std::cout << "Enter operands separated by spaces: ";
        std::cin >> op1 >> op2;
        std::cout << "Enter instruction character: ";
        std::cin >> ins;
        std::map<char, int> functionMap;
        //int (*add)(int, int) = _add;
        //int (*sub)(int, int) = _sub;
        functionMap['+'] = _add(op1,op2);
        functionMap['-'] = _sub(op1,op2);
        functionMap['*'] = _mul(op1,op2);
        functionMap['/'] = _div(op1,op2);
        std::cout << functionMap[ins];
        */

        //6.
        //char* s1 = new char[100];
        //std::cin.getline(s1); 
        //char* s2 = new char[100];
        //std::cin.getline(s2);
        //std::cout << s1 << std::endl;
        //std::cout << s2 << std::endl;
        //s1 = _strcat(s1, s2);
        //std::cout << s1 << std::endl;

        //8.
        //_strmove(str, 7);

        //for (int i = 0; i < 13; ++i) {
        //        std::cout << str[i];
        //}

        //delete [] str;
        //delete [] s1;
        //delete [] s2;
}

//1.
int binarySearchIterative(std::vector<int>& arr, int target) {
        int begin = 0;
        int end = arr.size() - 1;
        int mid = -1;
        while (begin <= end) {
                mid = (begin + end) / 2;
                if (target == arr[mid]) {
                        return mid;
                } 
                if (target < arr[mid]) {
                        end = mid - 1;
                } else {
                        begin = mid + 1;
                }
        }
        return mid;
}

int binarySearchRecursive(std::vector<int>& arr, int target, int begin, int end) {
        int mid = -1;
        if (begin <= end) {
                mid = (begin + end) / 2;
                if (target == arr[mid]) {
                        return mid;
                }
                if (target < arr[mid]) {
                        end = mid - 1;
                        return binarySearchRecursive(arr, target, begin, end);
                } else {
                        begin = mid + 1;
                        return binarySearchRecursive(arr, target, begin, end);
                }
        }
        return mid;
}

//2.
int _strlen(char* str) {
        int count = 0;
        while (str[count] != '\0') {
                ++count;
        }
        return count;
}

//3.
void reverse(char* str) {
        int length = _strlen(str);
        char tmp;
        for (int i = 0; i < length / 2; ++i) {
                tmp = str[i];
                str[i] = str[length - i];
                str[length - i] = tmp;
        }
}

//4.
void to_upper_case(char* str) {
        int length = _strlen(str);
        char tmp;
        for (int i = 0; i < length; ++i) {
                if (str[i] >= 97 && str[i] <= 122) {
                        str[i] -= 32;
                }
        }
}

//5.
int _add(int op1, int op2) {
        
        return op1 + op2;
}

int _sub(int op1, int op2) {
        return op1 - op2;
}

int _mul(int op1, int op2) {
        return op1 * op2;
}

int _div(int op1, int op2) {
        if (op2 != 0) {
                return op1 / op2;
        }
        std::cout << "C++ does not divide a number by zero, unlike JS or Python";
        return 0;
}

//6.
char* _strcat(char* s1, char* s2) {
        int size1 = _strlen(s1);
        int size2 = _strlen(s2);
        for (int i = 0; i < size2; ++i) {
                s1[size1 + i] = s2[i];
        }
        return s1;
}

//7.
void _strset(char* s, char ch) {
        int length = _strlen(s);
        for (int i = 0; i < length; ++i) {
                s[i] = ch;
        }
}

//8.
void _strmove(char* str, int n) {
        int length = _strlen(str);
        int count = n % length;
        char tmp;
        for (int i = 0; i < count; ++i) {
                tmp = str[length - 1];
                for (int j = length - 1; j > 0; --j) {
                        str[j] = str[j - 1];
                }
                str[0] = tmp;
        }
}

File: homework_06_09_22/homework.java
import java.util.*;

public class Homework {
        public static Scanner sc = new Scanner(System.in);

        static char[] toCharArray(String str, int length) {
                char[] arr = new char[length];
                for (int i = 0; i < str.length(); ++i) {
                        arr[i] = str.charAt(i);
                }
                arr[str.length()] = '\0';
                return arr;
        }

        public static void main(String argc[]) {
                //1.
                /*
                System.out.print("Enter array size: ");
                int size = sc.nextInt();
                System.out.print("Enter array elements: ");
                int[] arr = new int[size];
                for (int i = 0; i < size; ++i) {
                        arr[i] = sc.nextInt();
                }
                System.out.print("Enter target: ");
                int target = sc.nextInt();
                int begin = 0;
                int end = size - 1;
                //int res = binarySearchIterative(arr, size, target);
                int res = binarySearchRecursive(arr, size, target, begin, end);
                System.out.println("The index is: " + res);
                */

                //3. - //4.
                /*
                int n;
                System.out.print("Enter string length: ");
                n = sc.nextInt();
                char[] str = new char[n + 1];
                for (int i = 0; i < n ; ++i) {
                        str[i] = sc.next().charAt(0);
                }
                str[n] = '\0';
                System.out.println(str);
                //3
                //reverse(str);
                //4
                to_upper_case(str);
                System.out.println(str);
                */

                //5.
                /*
                System.out.print("Enter operands separated by space: ");
                Double op1 = sc.nextDouble();
                Double op2 = sc.nextDouble();
                System.out.print("Enter valid(+,-,*,/) instruction character: ");
                Character inst = sc.next().charAt(0);
                Map<Character, Double> map = new HashMap<Character, Double>();
                map.put('+', _add(op1, op2));
                map.put('-', _sub(op1, op2));
                map.put('*', _mul(op1, op2));
                map.put('/', _div(op1, op2));
                System.out.println(map.get(inst));
                */

                //6.
                /*
                String s1 = sc.nextLine();
                String s2 = sc.nextLine();
                int length = s1.length() + s2.length();
                char[] str1 = new char[length + 1];
                char[] str2 = new char[s2.length() + 1];
                str1 = toCharArray(s1, length + 1);
                str2 = toCharArray(s2, s2.length() + 1);
                _strcat(str1, str2);
                System.out.println(str1);
                */

                //7. - //8.
                /*
                String s = sc.nextLine();
                char[] str = new char[s.length() + 1];
                int n = sc.nextInt();
                str = toCharArray(s, s.length() + 1);
                //char ch = sc.next().charAt(0);
                //_strset(str, ch);
                _strmove(str, n);
                System.out.println(str);
                */
        }

        //1.
        static int binarySearchIterative(int[] arr,int size, int target) {
                int begin = 0;
                int end = size - 1;
                int mid = -1;
                while (begin <= end) {
                        mid = (begin + end) / 2;
                        if (arr[mid] == target) {
                                return mid;
                        }
                        if (arr[mid] < target) {
                                begin = mid + 1;
                        } else {
                                end = mid - 1;
                        }
                }
                return mid;
        }

        static int binarySearchRecursive(int arr[], int size, int target, int begin, int end) {
                int mid = -1;
                if (begin <= end) {
                        mid = (begin + end) / 2;
                        if (arr[mid] == target) {
                                return mid;
                        }
                        if (arr[mid] < target) {
                                begin = mid + 1;
                                return binarySearchRecursive(arr, size, target, begin, end);
                        } else {
                                end = mid - 1;
                                return binarySearchRecursive(arr, size, target, begin, end);
                        }
                }
                return mid;
        }
        

        //2.
        static void reverse(char[] str) {
                int length = _strlen(str);
                char tmp = ' ';
                for (int i = 0; i < length / 2; ++i) {
                        tmp = str[i];
                        str[i] = str[length - i - 1];
                        str[length - i - 1] = tmp;
                }
        }
        
        //3.
        static int _strlen(char[] str) {
                int count = 0;
                while (str[count] != '\0') {
                        ++count;
                }
                return count;
        }

        //4.
        static void to_upper_case(char[] str) {
                for (int i = 0; str[i] != '\0'; ++i) {
                        if (str[i] >= 97 && str[i] <= 122) {
                                str[i] -= 32;
                        }
                }
        }

        //5.
        static Double _add(Double op1, Double op2) {
                return op1 + op2;
        }

        static Double _sub(Double op1, Double op2) {
                return op1 - op2;
        }

        static Double _mul(Double op1, Double op2) {
                return op1 * op2;
        }

        static Double _div(Double op1, Double op2) {
                if (op2 != 0) {
                        return op1 / op2;
                }
                System.out.println("Division by zero!");
                return 0.0;
        }
        
        //6.
        static void _strcat(char[] str1, char[] str2) {
                int length1 = _strlen(str1);
                int length2 = _strlen(str2) - 1;
                for (int i = 0; i <= length2 ; ++i) {
                        str1[length1 + i] = str2[i];
                }      
        }
        
        //7.
        static void _strset(char[] str, char ch) {
                int length = _strlen(str);
                for (int i = 0; i < length; ++i) {
                        str[i] = ch;
                }
        }

        static void _strmove(char[] str, int n) {
                int length = _strlen(str);
                int count = n % (length);
                char tmp;
                for (int i = 0; i < count; ++i) {
                        tmp = str[length - 1];
                        for (int j = 0; j < length - 1; ++j) {
                                str[length - 1 - j] = str[length - 2 - j];
                        }
                        str[0] = tmp;
                }
        }
}
